syntheseus/cli/search.py

```diff
@@ -0,0 +1,310 @@
+        build_node_evaluator("policy")
+
+        alg_kwargs["bound_function"] = lookup_by_name(mcts_base, alg_kwargs["bound_function_class"])
+        del alg_kwargs["bound_function_class"]
+
+        alg = MolSetMCTS(**alg_kwargs)
+    else:
+        raise NotImplementedError(f"Unsupported search algorithm {config.search_algorithm}")
+
+    # Prepare the output directory
+    results_dir_top_level = Path(config.results_dir)
+    timestamp = datetime.datetime.now().isoformat(timespec="seconds")
+    results_dir_current_run = results_dir_top_level / f"{config.model_class.name}_{str(timestamp)}"
+
+    logger.info("Setup completed")
+    num_targets = len(search_targets)
+
+    all_stats: List[Dict[str, Any]] = []
+    for idx, smiles in enumerate(tqdm(search_targets)):
+        logger.info(f"Running search for target {smiles}")
+
+        if num_targets == 1:
+            results_dir = results_dir_current_run
+        else:
+            results_dir = results_dir_current_run / str(idx)
+
+        results_dir.mkdir(parents=True, exist_ok=True)
+        logger.info(f"Outputs will be saved under {results_dir}")
+
+        alg.reset()
+        output_graph, _ = alg.run_from_mol(Molecule(smiles))
+        logger.info(f"Finished search for target {smiles}")
+
+        # Time of first solution (rxn model calls)
+        for node in output_graph.nodes():
+            node.data["analysis_time"] = node.data["num_calls_rxn_model"]
+        soln_time_rxn_model_calls = get_first_solution_time(output_graph)
+
+        # Time of first solution (wallclock)
+        for node in output_graph.nodes():
+            node.data["analysis_time"] = (
+                node.creation_time - output_graph.root_node.creation_time
+            ).total_seconds()
+        soln_time_wallclock = get_first_solution_time(output_graph)
+
+        stats = {
+            "index": idx,
+            "smiles": smiles,
+            "rxn_model_calls_used": alg.reaction_model.num_calls(),
+            "num_nodes_in_final_tree": len(output_graph),
+            "soln_time_rxn_model_calls": soln_time_rxn_model_calls,
+            "soln_time_wallclock": soln_time_wallclock,
+        }
+
+        all_stats.append(stats)
+        logger.info(pformat(stats))
+
+        with open(results_dir / "stats.json", "wt") as f_stats:
+            f_stats.write(json.dumps(stats, indent=2))
+
+        if config.save_graph:
+            with open(results_dir / "graph.pkl", "wb") as f_graph:
+                pickle.dump(output_graph, f_graph)
+
+        if config.num_routes_to_plot > 0:
+            # Extract some synthesis routes in the order they were found
+            logger.info(f"Extracting up to {config.num_routes_to_plot} routes for analysis")
+
+            # TODO(kmaziarz): Add options to extract a diverse (or otherwise interesting) subset.
+            routes: Iterator = iter_routes_time_order(
+                output_graph, max_routes=config.num_routes_to_plot
+            )
```

Comment:
Rather than iterate over minimum cost routes, we could also randomly sample some routes. I have some code for this in another repo that I could add later.

This is just a suggestion though, for now let's merge it like this.